### DISCUSSION TITLE
Communication between PLC and PC

### DISCUSSION THREAD

**Discussion by Schadi0815** [gijzelaerr/python-snap7#500] [Q&A] [answered]
Hello,

I am trying to establish communication between a non-optimized DB in my PLC and a PC.
On the PC I am using the Python programming language with a Snap7 library.
The PUT/GET function is activated on the PLC.

Communication must be very fast (~100ms).
The PLC is a 1513F-1 PN with the following part number: 6ES7 513-1FL02-0AB0 and this firmware version: V2.9
The cycle time of the PLC is around 40-60 ms.

Unfortunately, the evaluation is not stable.
For example: 
We have physically mounted two triggers. One trigger goes to the PLC
The other trigger goes to a camera, which triggers if the trigger is interrupted. The images are retrieved from the PC via a Python script. This also works so far.

The PC counts up 1 for each photo and sends the value to the PLC. Unfortunately, not all values written by the PC arrive at the PLC. In other words: 1000 triggers sent to the PC --> 997 triggers arrive in the PLC
Does anyone have any experience as to why this could be and can they please help me?

**Comment by nikteliy** [comment]
Such behavior can have multiple reasons, and the python-snap7 library is far from being the first one. In any case, without details, it's difficult to suggest anything.

I would start by checking the PLC settings. For example, this may be related to the short cycle time. You might want to try increasing it.
![image](https://github.com/gijzelaerr/python-snap7/assets/52915342/ee87e375-8603-41a3-a028-97340b09afe0)

**Reply by Mr-OT** [reply]
I am making an OT Asset inventory by querrying the PLC IP via snap7 module and i am getting this SZL = Master S7-400 416-3 CPU 416-3 PN/DP Original Siemens Equipment SVPV6416951 CPU 416-3 PN/DP MC SVPWN107858 * Endmontage And also another 6ES7 315-2EH13-0AB0 6ES7 315-2EH13-0AB0 V Boot Loader A . But this return response is always missing the Version . What mistakes am i doing can you please tell me how can i make the plc to send me back the CPU firmware version back. I have all the info except the Firmware

Sample Querry 
def query(ip):
    client = try_connect(ip)
    if not client:
        return {"Status": "❌ Could not connect on any rack/slot"}

    try:
        id_bytes = read_szl_info(client, 0x0011, 0)
        fw_bytes = read_szl_info(client, 0x001C, 0)

        id_data = decode_szl_block(id_bytes, "ID")
        fw_data = decode_szl_block(fw_bytes, "FW")

        client.disconnect()

**Comment by gijzelaerr** [comment, marked as answer]
Closing as answered. The missing trigger pulses are likely a timing/PLC configuration issue, not a library bug.
